### PR TITLE
Support multi-part GCM mode encryption

### PR DESCRIPTION
**Pull request by balvisio**
@jethrogb @Goirad @jseyfried :Can you please review this? It AES-GCM encrytion support.

### CODE REVIEW

**Comment by balvisio** [comment]
@jseyfried

File: mbedtls/src/cipher/mod.rs [hunk 1]
```diff
@@ -124,6 +124,10 @@ impl<O: Operation, T: Type, S: State> Cipher<O, T, S> {
             None
         }
     }
+
+    pub fn mode(&self) -> raw::CipherMode {
```
Comment (mzohreva) [inline mbedtls/src/cipher/mod.rs:128]: rename this to `cipher_mode`

File: mbedtls/src/cipher/raw/serde.rs [hunk 2]
```diff
@@ -294,6 +364,7 @@ unsafe impl BytesSerde for cipher_context_t {}
 unsafe impl BytesSerde for aes_context {}
 unsafe impl BytesSerde for des_context {}
 unsafe impl BytesSerde for des3_context {}
+unsafe impl BytesSerde for gcm_context {}
 
 // If the C API changes, the serde implementation needs to be reviewed for correctness.
```
Comment (mzohreva) [inline mbedtls/src/cipher/raw/serde.rs:370]: Please add the following at the end of this file:
```rust
unsafe fn _check_gcm_context_size(ctx: gcm_context) -> [u8; 424] {
    ::core::mem::transmute(ctx)
}
```

File: mbedtls/src/cipher/raw/serde.rs [hunk 3]
```diff
@@ -87,6 +90,35 @@ impl<Op: Operation> Serialize for Cipher<Op, Traditional, CipherData> {
...
+                        key_bit_len: inner_ctx_key_bit_len,
+                        context: Bytes(inner_ctx),
+                        algorithm_ctx: AlgorithmContext::Aes(Bytes(inner_aes_context)),
+                    };
+
+                    AlgorithmContext::Gcm {
+                        context: Bytes(*(cipher_context.cipher_ctx as *const gcm_context)),
+                        inner_cipher_ctx: Box::new(inner_cipher_ctx)
+                    }
+                },
```
Comment (mzohreva) [inline mbedtls/src/cipher/raw/serde.rs:121]: This seems like a mine field and I don't feel qualified to review it. @jethrogb please review this part as well as the deserialize part.

File: mbedtls/src/cipher/raw/serde.rs [hunk 4]
```diff
@@ -87,6 +90,35 @@ impl<Op: Operation> Serialize for Cipher<Op, Traditional, CipherData> {
                 | (CIPHER_ID_3DES, MODE_CFB) => AlgorithmContext::Des3(Bytes(
                     *(cipher_context.cipher_ctx as *const des3_context),
                 )),
+                (CIPHER_ID_AES, MODE_GCM) => {
+                    let gcm_context = *(cipher_context.cipher_ctx as *const gcm_context);
+
+                    let mut inner_ctx = gcm_context.cipher_ctx;
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:96]: Instead of manually serializing the inner context, you should just recursively call this logic. Same thing for deserialization.

File: mbedtls/src/cipher/raw/serde.rs [hunk 5]
```diff
@@ -117,8 +71,75 @@ impl<Op: Operation> Serialize for Cipher<Op, Traditional, CipherData> {
     }
 }
 
-impl<'de, Op: Operation> Deserialize<'de> for Cipher<Op, Traditional, CipherData> {
-    fn deserialize<D>(d: D) -> Result<Cipher<Op, Traditional, CipherData>, D::Error>
+unsafe fn serialize_raw_cipher<S: Serializer>(mut cipher_context: cipher_context_t)
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:74]: I don't understand why this is taking type parameter S?

Comment (jethrogb) [reply mbedtls/src/cipher/raw/serde.rs:74]: Oh to be able to return an error.

File: mbedtls/src/cipher/raw/serde.rs [hunk 6]
```diff
@@ -167,36 +189,7 @@ impl<'de, Op: Operation> Deserialize<'de> for Cipher<Op, Traditional, CipherData
...
-                }
-            };
-
-            cipher_context.key_bitlen = raw.context.0.key_bitlen;
-            cipher_context.operation = raw.context.0.operation;
-            cipher_context.unprocessed_data = raw.context.0.unprocessed_data;
-            cipher_context.unprocessed_len = raw.context.0.unprocessed_len;
-            cipher_context.iv = raw.context.0.iv;
-            cipher_context.iv_size = raw.context.0.iv_size;
+            deserialize_raw_cipher::<D>(raw, cipher_context)?;
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:192]: Can you move lines 167-188 into deserialize_raw_cipher as well so you don't have to repeat cipher setup for GCM internals?

File: mbedtls/src/cipher/raw/serde.rs [hunk 7]
```diff
@@ -209,6 +202,63 @@ impl<'de, Op: Operation> Deserialize<'de> for Cipher<Op, Traditional, CipherData
     }
 }
 
+unsafe fn deserialize_raw_cipher<'de, D>(raw: SavedRawCipher, cipher_context: &mut cipher_context_t)
+    -> Result<(), D::Error>
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:206]: Return `Result<(), (&'static str, &'static str)>` instead so you don't have to pass `D` around.

File: mbedtls/src/cipher/raw/serde.rs [hunk 8]
```diff
@@ -117,8 +71,75 @@ impl<Op: Operation> Serialize for Cipher<Op, Traditional, CipherData> {
     }
 }
 
-impl<'de, Op: Operation> Deserialize<'de> for Cipher<Op, Traditional, CipherData> {
-    fn deserialize<D>(d: D) -> Result<Cipher<Op, Traditional, CipherData>, D::Error>
+unsafe fn serialize_raw_cipher<S: Serializer>(mut cipher_context: cipher_context_t)
+    -> Result<SavedRawCipher, S::Error> {
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:75]: Return `Result<SavedRawCipher, &'static str>` instead so you don't have to pass `S` around.

File: mbedtls/src/cipher/raw/serde.rs [hunk 9]
```diff
@@ -312,3 +363,5 @@ unsafe fn _check_des_context_size(ctx: des_context) -> [u8; 128] {
 unsafe fn _check_des3_context_size(ctx: des3_context) -> [u8; 384] {
     ::core::mem::transmute(ctx)
 }
+
+unsafe fn _check_gcm_context_size(ctx: gcm_context) -> [u8; 424] { ::core::mem::transmute(ctx) }
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:367]: missing newline at end of file

File: mbedtls/src/cipher/raw/serde.rs [hunk 10]
```diff
@@ -46,68 +46,23 @@ enum AlgorithmContext {
...
-            SavedRawCipher {
-                cipher_id: cipher_id,
-                cipher_mode: cipher_mode,
-                key_bit_len: key_bit_len,
-                context: Bytes(cipher_context),
-                algorithm_ctx: algorithm_ctx,
-            }
+            let cipher_context = self.raw_cipher.inner;
+            serialize_raw_cipher(cipher_context)
+                .map_err(|e| ser::Error::custom(e))?
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:65]: Nit: can write this as `.map_err(ser::Error::custom)`

File: mbedtls/src/cipher/raw/serde.rs [hunk 11]
```diff
@@ -117,8 +72,73 @@ impl<Op: Operation> Serialize for Cipher<Op, Traditional, CipherData> {
...
+        )),
+        (CIPHER_ID_AES, MODE_GCM) => {
+            let gcm_context = *(cipher_context.cipher_ctx as *const gcm_context);
+
+            let inner_ctx = gcm_context.cipher_ctx;
+
+            let inner_saved_cipher = serialize_raw_cipher(inner_ctx)?;
+
+            AlgorithmContext::Gcm {
+                context: Bytes(*(cipher_context.cipher_ctx as *const gcm_context)),
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:115]: `Bytes(gcm_context)`

File: mbedtls/src/cipher/raw/serde.rs [hunk 12]
```diff
@@ -137,76 +157,86 @@ impl<'de, Op: Operation> Deserialize<'de> for Cipher<Op, Traditional, CipherData
...
+        (CIPHER_ID_AES, AlgorithmContext::Gcm {
+            context,
+            inner_cipher_ctx
+        }) => {
+            let ret_gcm_context = cipher_context.cipher_ctx as *mut gcm_context;
+            *ret_gcm_context = context.0;
+
+            let raw_cipher = deserialize_raw_cipher(*inner_cipher_ctx, raw::CipherPadding::None)?;
+            (*ret_gcm_context).cipher_ctx = raw_cipher.into_inner();
+        }
```
Comment (jethrogb) [inline mbedtls/src/cipher/raw/serde.rs:226]: Bring this branch in line with the way the others are written:
```rust
        (CIPHER_ID_AES, AlgorithmContext::Gcm {
            context: Bytes(gcm_ctx),
            inner_cipher_ctx
        }) => {
            let inner_raw_cipher = deserialize_raw_cipher(*inner_cipher_ctx, raw::CipherPadding::None)?;
            gcm_ctx.cipher_ctx = inner_raw_cipher.into_inner();

            *(cipher_context.cipher_ctx as *mut gcm_context) = gcm_ctx
        }
```

File: mbedtls/src/cipher/mod.rs [hunk 13]
```diff
@@ -17,13 +17,15 @@ pub trait Operation: Sized {
     fn is_encrypt() -> bool;
 }
 
+#[derive(Serialize, Deserialize)]
```
Comment (mzohreva) [inline mbedtls/src/cipher/mod.rs:20]: Why is this needed?

The `Encryption` and `Decryption` types are indistinguishable in serialized form, e.g. in JSON they both serialize to `{}`.
